net_transport.go

```diff
@@ -380,14 +413,20 @@ func (n *NetworkTransport) returnConn(conn *netConn) {
 // AppendEntriesPipeline returns an interface that can be used to pipeline
 // AppendEntries requests.
 func (n *NetworkTransport) AppendEntriesPipeline(id ServerID, target ServerAddress) (AppendPipeline, error) {
+	if n.maxInFlight < 2 {
+		// Pipelining is disabled since no more than one request can be outstanding
+		// at once. Skip the whole code path and use synchronous requests.
+		return nil, ErrPipelineReplicationNotSupported
+	}
+
```

Comment:
What is the implication of making a pipeline of size 1 instead of retuning an error in that use case? With this, if we decide to thread this config up to Consul would we need to add extra logic to call the right API?

Reply:
A pipeline of size one doesn't make any sense with our current implementation and will panic.

This is because we only block on the buffer _after_ sending the next request, and because the first request sent is immediately consumed by the receiving go routine. That means even with a totally buffer size of zero you still allow 2 requests to be in flight: one because the receiving goroutine immediately consumes and unblocks the buffer while it waits, and the second because we send it before pushing to inflight where we block waiting for the first response.

We _could_ change the behaviour of the pipeline to push to the channel first before sending so that a zero-buffered chan would correspond to `MaxInFlight = 1`... But I don't see much benefit to that - that would just make it ambigous whether `MaxInFlight = 1` mean "no pipeline at all" or "use pipelining code path for parallel decoding/encoding, but don't allow more than one outstanding request. In theory that still has some potential performance benefit but I was not able to measure one so I don't think it's worth adding the extra complexity/risk by changing how pipelining works and making there be even more subtle config options.

Does that make sense?

The flip side to this is that it might in some ways be simpler to make the config hide the complexity less by exposing a separate `DisablePipelining` and `PipelineBufferSize`, but then operators have to work out which thing they want and we still have to document all the unintuitive stuff like "Buffer size of 0 actually allows 2 requests to be in flight". I don't have a strong opinion but choose to lean towards making the config map more directly to the behaviour it enabled rather than obscure implementation details.

Reply:
The point I was trying to make is mostly that from UX point of you I expect a pipeline size of 1 or 0 (depending on how we design it) to work as if I call `AppendEntries()` without any pipelining. I guess we already missed that opportunity when we made the choice to make separate APIs for the pipelining option way before this PR. The user, in that case raft, need to deal with making the right choice which is already in place.

I agree hard failing, by panic is a reasonable choice to avoid misconfiguration here.